### PR TITLE
Make SparseTensor type first class citizen.

### PR DESCRIPTION
**Pull request by yuslepukhin**
Adjust tests, checker and shape inference.

Signed-off-by: Dmitri Smirnov <dmitrism@microsoft.com>

### CODE REVIEW

**Comment by askhade** [comment]
IR_Version needs to be updated too right?

File: onnx/shape_inference/implementation.cc [hunk 1]
```diff
@@ -77,6 +87,39 @@ void checkShapesAndTypes(const TypeProto_Tensor& inferredType, const TypeProto_T
   }
 }
 
+void checkShapesAndTypes(const TypeProto_SparseTensor& inferredType, const TypeProto_SparseTensor& existingType) {
```
Comment (askhade) [inline onnx/shape_inference/implementation.cc:90]: Can this be simplified by using a template?

File: onnx/shape_inference/implementation.cc [hunk 2]
```diff
@@ -175,13 +251,14 @@ static void InferShapesImpl(
     }
   }
 
-  std::unordered_map<std::string, const TensorProto*> inputDataByName;
   // save for free memory
-  std::vector<TypeProto*> initializerTypeList;
+  std::vector<std::unique_ptr<TypeProto>> initializerTypeList;
```
Comment (askhade) [inline onnx/shape_inference/implementation.cc:255]: do we even need a unique_ptr here? Can the vector simply hold the TypeProto object?

**Comment by gramalingam** [comment]
A generic question: is this the same as the closed PR, or is there any significant difference in this PR?

File: onnx/defs/shape_inference.h [hunk 3]
```diff
@@ -154,131 +141,141 @@ multiplyDims(const TensorShapeProto& shape, int from, int upto_exclusive) {
   return dim;
 }
 
-void propagateElemTypeWithValidation(
-    const TypeProto* input_type,
-    TypeProto* output_type);
+inline int32_t getTensorElementType(const TypeProto& type) {
```
Comment (gramalingam) [inline onnx/defs/shape_inference.h:144]: (General comment, not specific to this PR). I wonder if this include file is worth splitting into an include-file and implementation-file (in a separate PR, doesn't have to be in this). It is growing big, and it is better to reduce unnecessary implementation-dependence some users might acquire.

**Comment by yuslepukhin** [comment]
There should be no difference as this was created to deal with DCO issues.

---
In reply to: [814347121](https://github.com/onnx/onnx/pull/3398#issuecomment-814347121) [](ancestors = 814347121)

**Comment by yuslepukhin** [comment]
Will do.

---
In reply to: [814311059](https://github.com/onnx/onnx/pull/3398#issuecomment-814311059) [](ancestors = 814311059)

File: onnx/defs/shape_inference.h (hunk 3 above)
Comment (yuslepukhin) [reply onnx/defs/shape_inference.h:144]: I was thinking along the same lines.

File: onnx/shape_inference/implementation.cc (hunk 1 above)
Comment (yuslepukhin) [reply onnx/shape_inference/implementation.cc:90]: Will do.

**Comment by yuslepukhin** [comment]
> IR_Version needs to be updated too right?

Will do

File: onnx/defs/shape_inference.h (hunk 3 above)
Comment (askhade) [reply onnx/defs/shape_inference.h:144]: I also agree

Comment (yuslepukhin) [reply onnx/defs/shape_inference.h:144]: I think this would fall under refactoring category and we better manage it under a separate PR.

**Review by jcwchen — COMMENTED** [review]
The CI is failing because https://github.com/onnx/onnx/blob/master/onnx/examples/make_model.ipynb checks ir_version as an old one (7).

Perhaps we can simply get rid of this line: `print('The ir_version in model: {}\n'.format(model_def.ir_version))`? Otherwise every time ONNX bumps ir_version, `make_model.ipynb` needs to be updated as well...

File: onnx/examples/make_model.ipynb [hunk 4]
```diff
@@ -9,7 +9,6 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "The ir_version in model: 7\n",
       "\n",
```
Comment (jcwchen) [inline onnx/examples/make_model.ipynb:12]: I think L13 needs to be removed as well. The check here is quite unfriendly for people who want to update it... I will try to improve it in another PR. Thank you for the effort.

**Review by gramalingam — APPROVED** [review]

File: onnx/defs/shape_inference.h [hunk 5]
```diff
@@ -61,6 +60,9 @@ struct InferenceContext {
   virtual TypeProto* getOutputType(size_t index) = 0;
   virtual GraphInferencer* getGraphAttributeInferencer(const std::string& attribute_name) = 0;
   virtual ~InferenceContext() {}
+  /// XXX: Putting at the end of vtable in case it matters?
+  /// XXX: Do we need default implementation here?
```
Comment (askhade) [inline onnx/defs/shape_inference.h:64]: no we dont... you can remove this comment

File: onnx/onnx.in.proto [hunk 6]
```diff
@@ -100,7 +100,12 @@ enum Version {
   //   method and training algorithm. The execution of TrainingInfoProto
   //   can modify the values of mutable variables.
   // - Implicitly add inference graph into each TrainingInfoProto's algorithm.
-  IR_VERSION = 0x0000000000000007;
+  IR_VERSION_2020_5_8 = 0x0000000000000007;
+
+  // IR VERSION 8 published on <TBD>
```
Comment (askhade) [inline onnx/onnx.in.proto:105]: may be we can change this to published with ONNX release 1.10? This way no one will forget to update the TBD

**Review by askhade — APPROVED** [review]
Add 2 nit comments other than that looks good to me. Thanks for waiting till ONNX 1.9 release

**Review by jcwchen — APPROVED** [review]

File: onnx/defs/shape_inference.h (hunk 5 above)
Comment (gramalingam) [reply onnx/defs/shape_inference.h:64]: I think a default implementation does help users who have an implementation of InferenceContext use this version without modifying their code (which seems reasonable if they do not want to handle sparse data yet).

**Pull request merged**